rust/arrow/src/compute/kernels/take.rs

```diff
@@ -0,0 +1,519 @@
+pub struct TakeOptions {
+    /// perform bounds check before taking
+    pub check_bounds: bool,
+}
+
+impl Default for TakeOptions {
+    fn default() -> Self {
+        Self {
+            check_bounds: false,
+        }
+    }
+}
+
+/// `take` implementation for numeric arrays
+fn take_numeric<T>(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef>
+where
+    T: ArrowNumericType,
+{
+    let mut builder = PrimitiveBuilder::<T>::new(index.len());
+    let a = array.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append_null()?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_valid(ix) {
+                builder.append_value(a.value(ix))?;
+            } else {
+                builder.append_null()?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for binary arrays
+fn take_binary(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
+    let mut builder = BinaryBuilder::new(index.len());
+    let a = array.as_any().downcast_ref::<BinaryArray>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append(false)?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_null(ix) {
+                builder.append(false)?;
+            } else {
+                builder.append_values(a.value(ix))?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for boolean arrays
+fn take_bool(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
+    let mut builder = BooleanBuilder::new(index.len());
+    let a = array.as_any().downcast_ref::<BooleanArray>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append_null()?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_null(ix) {
+                builder.append_null()?;
+            } else {
+                builder.append_value(a.value(ix))?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for list arrays
+///
+/// Calculates the index and indexed offset for the inner array,
+/// applying `take` on the inner array, then reconstructing a list array
+/// with the indexed offsets
+fn take_list(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
+    let list: &ListArray = array.as_any().downcast_ref::<ListArray>().unwrap();
+    let (indices, offsets) = take_index_from_list(array, index);
+    let taken = take(&list.values(), &indices, None)?;
+    let value_offsets = Buffer::from(offsets[..].to_byte_slice());
+    let list_data = ArrayData::new(
+        list.data_type().clone(),
+        index.len(),
+        Some(index.null_count()),
+        taken.data().null_bitmap().clone().map(|bitmap| bitmap.bits),
```

Comment:
Is this correct? seems it is using value array's null count.

Reply:
Hi @sunchao, this was quite difficult to deal with; but I've fixed it. The index's null count was only correct if the `values` array had no null counts. I now recompute the null count from the offsets, where an offset of `vec![0,1,1,2,2]` contains 2 null values as 1 and 2 repeat. I'm getting ready to push the latest changes.